File: 2_add_two_numbers/main.go
package main

import "fmt"

// Definition for singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{}
	current := head
	carry := 0
	for l1 != nil || l2 != nil || carry > 0 {
		x := 0
		y := 0

		if l1 != nil {
			x = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			y = l2.Val
			l2 = l2.Next
		}
		result := x + y + carry
		newNode := &ListNode{
			Val: (result % 10),
		}
		carry = (result / 10)

		current.Next = newNode
		current = current.Next
	}
	return head
}

func createListWithArray(arr []int) *ListNode {
    if len(arr) == 0 {
        return nil
    }
    head := &ListNode{Val: arr[0]}
    current := head
    for _, val := range arr[1:] {
        new_node := &ListNode{Val: val}
        current.Next = new_node
        current = current.Next
    }
    return head
}

func main() {
	l1_arr := []int{2, 4, 3}
	l1 := createListWithArray(l1_arr)
	l2_arr := []int{5, 6, 4}
	l2 := createListWithArray(l2_arr)
	teste := addTwoNumbers(l1, l2)
	fmt.Println(teste)
}

File: 2_add_two_numbers/main.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode(val=None)
        current = head
        a = 0
        while l1 or l2 or a:
            x = l1.val if l1 is not None else 0
            y = l2.val if l2 is not None else 0
            result = x + y + a
            a = result // 10

            if current.val is None:
                current.val = result % 10
            else:
                current.next = ListNode(result % 10)
                current = current.next

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return head

def create_linked_list(arr):
    dummy = ListNode()
    current = dummy

    for num in arr:
        current.next = ListNode(num)
        current = current.next

    return dummy.next

l1 = create_linked_list([9,9,9,9,9,9,9])
l2 = create_linked_list([9,9,9,9])

solution = Solution()
result = solution.addTwoNumbers(l1, l2)

while result:
    print(result.val, end=" ")
    result = result.next